src/client/Microsoft.Identity.Client/Platforms/Android/Broker/AndroidContentProviderBroker.cs

```diff
@@ -0,0 +1,349 @@
+            {
+                _logger.ErrorPiiWithPrefix(ex, "Android broker silent invocation failed. ");
+                _brokerHelper.HandleBrokerOperationError(ex);
+                throw;
+            }
+        }
+
+        private async Task<MsalTokenResponse> AcquireTokenSilentViaBrokerAsync(BrokerRequest brokerRequest)
+        {
+            using (_logger.LogMethodDuration())
+            {
+                _logger.Verbose("User is specified for silent token request. Starting silent Android broker request. ");
+
+                brokerRequest = await UpdateRequestWithAccountInfoAsync(brokerRequest).ConfigureAwait(false);
+
+                string silentResult = await PerformAcquireTokenSilentFromBrokerAsync(brokerRequest).ConfigureAwait(false);
+                return _brokerHelper.HandleSilentAuthenticationResult(silentResult, brokerRequest.CorrelationId);
+            }
+        }
+
+        /// <summary>
+        /// This method is only used for Silent authentication requests so that we can check to see if an account exists in the account manager before
+        /// sending the silent request to the broker. 
+        /// </summary>
+        public async Task<BrokerRequest> UpdateRequestWithAccountInfoAsync(BrokerRequest brokerRequest)
+        {
+            var accountData = await GetBrokerAccountDataAsync(brokerRequest).ConfigureAwait(false);
+
+            return _brokerHelper.UpdateBrokerRequestWithAccountData(accountData, brokerRequest);
+        }
+
+        public async Task<string> PerformAcquireTokenSilentFromBrokerAsync(BrokerRequest brokerRequest)
+        {
+            Bundle silentOperationBundle = _brokerHelper.CreateSilentBrokerBundle(brokerRequest);
+            var OperationBundleJSON = _brokerHelper.SearializeBundleToJSON(silentOperationBundle);
+            var SilentOperationBundle = await PerformContentResolverOperationAsync(ContentResolverOperation.acquireTokenSilent, OperationBundleJSON).ConfigureAwait(false);
+            return _brokerHelper.GetSilentResultFromBundle(SilentOperationBundle);
+        }
+
+        public async Task<IEnumerable<IAccount>> GetAccountsAsync(string clientID, string redirectUri)
+        {
+            using (_logger.LogMethodDuration())
+            {
+                if (!IsBrokerInstalledAndInvokable())
+                {
+                    _logger.Warning("Android broker is either not installed or is not reachable so no accounts will be returned. ");
+                    return new List<IAccount>();
+                }
+
+                return await PerformGetAccountsAsync(clientID, redirectUri).ConfigureAwait(false);
+            }
+        }
+
+        private async Task<IEnumerable<IAccount>> PerformGetAccountsAsync(string clientID, string redirectUri)
+        {
+            BrokerRequest brokerRequest = new BrokerRequest() { ClientId = clientID, RedirectUri = new Uri(redirectUri) };
+
+            try
+            {
+                InitiateBrokerHandshakeAsync();
```

Comment:
Should this be cached? If it worked once, should we do it over and over?

Reply:
It is possible to cache. i spoke with the android team adn they provided guidance on how to do this. I created a seperate issue here for this
https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/2390

Reply:
i asked this in the accountManager class...we definitely need to cache it. i would say that should happen before we ship this. let's discuss.

---
In reply to: [569855972](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/2380#discussion_r569855972) [](ancestors = 569855972)

Reply:
also needs to be awaited.

---
In reply to: [570655066](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/2380#discussion_r570655066) [](ancestors = 570655066,569855972)